packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+        // Only create and push new attestation if we haven't hit max
+        if (currIndex < maxAttestations) {
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation memory attestation = identifierToAddresses[identifier][trustedIssuers[i]][j];
+          if (!revokedSigners[attestation.signer]) {
+            accounts[currIndex] = attestation.account;
+            issuedOns[currIndex] = attestation.issuedOn;
+            signers[currIndex] = attestation.signer;
+            currIndex = currIndex.add(1);
+          }
+        } else {
+          break;
+        }
+      }
+    }
+
+    // Trim returned structs if necessary
+    if (currIndex < maxAttestations) {
+      address[] memory trimmedAccounts = new address[](currIndex);
+      uint256[] memory trimmedIssuedOns = new uint256[](currIndex);
+      address[] memory trimmedSigners = new address[](currIndex);
+
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedAccounts[i] = accounts[i];
+        trimmedIssuedOns[i] = issuedOns[i];
+        trimmedSigners[i] = signers[i];
+      }
+      return (trimmedAccounts, trimmedIssuedOns, trimmedSigners);
+    } else {
+      return (accounts, issuedOns, signers);
+    }
+  }
+
+  /**
+   * @notice Returns identifiers mapped (by unrevoked signers) to a
+   *   given account address by a list of trusted issuers
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @param maxIdentifiers Limit the number of identifiers that will be returned
+   * @return Array (length <= maxIdentifiers) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of identifiers per issuer
+
+  function lookupIdentifiersByAddress(
+    address account,
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
+  ) public view returns (bytes32[] memory) {
+    // Same as for the other lookup, preallocate and then trim for now
+    uint256 currIndex = 0;
+    bytes32[] memory identifiers = new bytes32[](maxIdentifiers);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      address trustedIssuer = trustedIssuers[i];
+      uint256 numIssuersForAddress = addressToIdentifiers[account][trustedIssuer].length;
+      for (uint256 j = 0; j < numIssuersForAddress; j = j.add(1)) {
+        // Iterate through the list of identifiers
+        if (currIndex < maxIdentifiers) {
+          bytes32 identifier = addressToIdentifiers[account][trustedIssuer][j];
+          // Check if the mapping was produced by a revoked signer
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation[] memory attestationsForIssuer = identifierToAddresses[identifier][trustedIssuer];
```

Comment:
As mentioned above, I think these should probably stay in storage unless we need to copy their data over to return

Reply:
+1. Using the `memory` modifier results in copying the full array into memory at this line via `SLOAD`. Because we may not iterate through the full array (we break if we find the account early) there may not be a need to `SLOAD` the entire array, it _may_ be more efficient to make this array as `storage` such that it is handled by storage reference and only loaded as needed in the loop. (Caveat here is that I am not a Solidity optimization expert).

Additionally note that this is making a tradeoff of execution efficiency for storage efficiency by not storing the signer and timestamp information in the reverse mapping. I think this is the right trade-off, but possible worth adding a comment that we are making it.

Reply:
Closing the loop here + since this is outdated code at this point: benchmarked the gas efficiency of this for lookups previously and settled on what seemed like the best compromise for the expected use case (not a ton of attestations per issuer/identifier). We can do more fine-tuning and benchmarking of this type after most of the PR's general function is finalized!